Remove methods that start with an underscore from at::Tensor

goldsborough opened the pull request:
This PR cleans up the `at::Tensor` class by removing all methods that start with an underscore in favor of functions in the `at::` namespace. This greatly cleans up the `Tensor` class and makes it clearer what is the public and non-public API.

For this I changed `native_functions.yaml` and `Declarations.cwrap` to make all underscore methods `variant: function` (or add such a statement to begin with), and then fixed all code locations using the underscore methods.

@ezyang @colesbury @gchanan

goldsborough committed:
Removed all _ functions except local_scalar from C++ tensor

goldsborough committed:
Remove Tensor::_local_scalar

goldsborough committed:
Missed _fused_dropout

goldsborough committed:
Add check to function_wrapper.py

zou3519 commented on torch/optim/sparse_adam.py:
```diff
@@ -66,8 +66,8 @@ def step(self, closure=None):
                 grad = grad.coalesce()  # the update is non-linear so indices must be unique
-                grad_indices = grad._indices()
-                grad_values = grad._values()
+                grad_indices = torch._indices(grad)
```
I think we should leave _indices and _values alone; sparse users use these fields a lot but we haven't changed the name from _* to something that doesn't have an _.

goldsborough replied:
Can we give them non-underscore names and make this a breaking change? This is certainly something we should fix and we do have a release upcoming, which gives us an opportunity.

What do you think? Also @soumith @SsnL

zou3519 replied:
cc @weiyangfb and @li-roy  for sparse things.

I feel uncomfortable breaking user code in one release; we should deprecate with warning and then break their code in a future release. On the other hand, idk how many people actually use sparse and we do promise no guarantees on our sparse API so...

weiyangfb replied:
I am fine with non-underscore names and give deprecate warning at the current underscored APIs

SsnL replied:
I'd argue that we should keep them.  For two reasons:

1. `_indices` and `_values` are really the properties/attributes of sparse tensors, so `torch._indices(tensor)` makes much less sense than `tensor._indices()`, at least to me. 

2. BC reasons as @zou3519 mentioned. Given how few sparse operations we support, many users of sparse tensors probably implemented their own sparse operations with `_indices` and `_values`.

FWIW, I am preparing a patch that adds `get_values` and `get_indices` methods, and those should be the public APIs, but they have subtle differences with `_values` and `_indices`, which make them much less useful in implementing custom sparse operations.

goldsborough replied:
What is we put them as methods on `Tensor` from within `torch/tensor.py`? As in, we'd add:

```
def _indices(self):
    return torch._indices(self)
```
That would preserve current users but keep the actual Tensor API clean from C++ side. I would still encourage you to come up with a deprecation strategy because calling an underscore method sounds/looks wrong either way, but that's up to you guys then. If there's any strong reason for keeping the method in C++ too we can make an exception for `_indices` and `_values`, but it sounds like a Python-only solution would work here?

zou3519 replied:
Many of the sparse methods implemented in aten use and mutate tensor._indices() / tensor._values(). It's not very nice looking at at::_indices(tensor) but I suppose as devs we can eat the notation.

goldsborough replied:
Ok why don't we (1) remove _indices/_values from Python (2) add `indices`/`values` (no underscore) (3) use `tensor.indices` in C++ (4) deprecate `_indices` in Python in favor of `indices`? Or some selection of these 4?

weiyangfb replied:
I agree with @SsnL and @zou3519 , `_indices` and `_values` are fields of sparse tensor, so it is unintuitive to call `at::_indices(tensor)`. I am also fine with `tensor.indices()` and `tensor.values()` and deprecate `_*` methods. btw, @SsnL are `get_indices` and `get_values` going to do same thing as `_indices` and `_values`?

SsnL replied:
Even in Python people can see/use these methods prefixed by underscores, and it hasn't caused us troubles. Why is this a problem for c++ API?

goldsborough replied:
It's not about trouble, it's just that we'd like the documentation for [at::Tensor](https://pytorch.org/cppdocs/api/structat_1_1_tensor.html#exhale-struct-structat-1-1-tensor) which is generated from the build/aten/src/ATen/Tensor.h file to be the exact public API we want users to use. In our Python docs we also filter out all the underscore methods except those three sparse "properties" https://imgur.com/a/Si2g5W5. I can make exceptions for those three, but just thinking about it, we should come up with a better API than an underscore method. It's not Pythonic at all!

SsnL replied:
My main concern is not about doc/public API, but that
1. They are really really attributes of sparse tensors! So it makes very little sense to say `at::indices(tensor)` (or`_indices`).
2. They are really really useful in writing sparse ops! 
So I think one reasonable solution is to provide some new better-named way to access these fields, so our c++ devs/users can use thoseto write the kernels. 

Fortunately this is exactly what you new commit does! :D although I would name them as `get_raw_values` and `get_raw_indices`, to indicate that they are unsafe and should usually be avoided.

cc @ezyang if he has opinions on any of these.

ezyang replied:
I believe the original reason why indices and values were prefixed with an underscore was because if you are working directly with `_values`, you have to be very careful about whether or not the sparse tensor is coalesced or not. Additionally, you can't differentiate through them, because `_indices` is defined to alias with the original tensor, but you can't do that if you have to wrap it in a new Variable to track the autograd (you also have differentiability problems if you allow returning duplicate entries!). @SsnL's pending patch closes these loopholes and gives you something that is semantically sound, but is less efficient than you would otherwise expect.

goldsborough committed:
Fix some function calls in CUDA and Python

SsnL commented on aten/src/ATen/cuda/CUDAApplyUtils.cuh:
```diff
@@ -620,23 +620,23 @@ bool CUDA_tensor_apply3(at::Tensor a,
     // instead, it will recursively try and invoke ourselves to make
     // oldA contiguous.
-    oldA._copy_ignoring_overlaps_(a);
+    at::_copy_ignoring_overlaps_(oldA, a);
```
This is breaking our convention. I don't think we've ever had inplace operations that are not methods. If you really want to change this, this should be `*_out`.

goldsborough replied:
`_copy_ignoring_overlaps_out` is fine with me

goldsborough committed:
_copy_ignoring_overlaps_ -> _copy_ignoring_overlaps_out

goldsborough committed:
Reintroduce tensor.{indices,values/nnz} and create Python wrappers

goldsborough commented:
Ok so I:
1. Renamed `{_values,_indices,_nnz}` to `{values,indices,nnz}` (no underscore) in C++ and expose those in Python
2. Created wrappers for `{_values,_indices,_nnz}` on torch/tensor.py to preserve BC in Python
3. Rewrote C++ code to use `.values()` instead of `.indices()`

I think this should make everyone happy @zou3519 @SsnL @weiyangfb?

ezyang commented:
> Renamed `{_values,_indices,_nnz}` to `{values,indices,nnz}` (no underscore) in C++ and expose those in Python

No, please don't do this. I'll quote the current sparse documentation:

> Second, some operators will produce different values depending on whether or not they are coalesced or not (e.g., torch.sparse.FloatTensor._values() and torch.sparse.FloatTensor._indices(), as well as torch.Tensor._sparse_mask()). These operators are prefixed by an underscore to indicate that they reveal internal implementation details and should be used with care, since code that works with coalesced sparse tensors may not work with uncoalesced sparse tensors; generally speaking, it is safest to explicitly coalesce before working with these operators.

https://pytorch.org/docs/stable/sparse.html

ezyang commented:
@goldsborough Is your goal to *unconditionally* remove all underscore-prefixed methods from Tensor, or are you OK with exceptions? `_indices` and `_values` is something of a live wire in the current sparse tensor API, and can't really be adequately resolved without some careful documentation and addition of new methods. I actually do think that something like `_indices` or `raw_indices` should be in the steady state sparse API, but they  should be clear that they have to be used with care.

goldsborough commented:
@ezyang I'm ok with exceptions, I just thought we might as well fix the API since public underscore fields seem wrong either way. But maybe that's a discussion for another PR. I'll leave them be in this one :)

goldsborough committed:
Make exceptions for _indices, _values, _nnz

ezyang reviewed (approved):
I just realized this is going to intersect with https://github.com/pytorch/pytorch/pull/11247 ; both diffs de-method-ize some functions where methods don't make sense (but yours is more aggressive) (I didn't realize this because the latter diff is doing something else: changing interpretation of missing variants in native). Do you want to fold that patch in, or should I just rebase on top? Or should I land mine first?

goldsborough commented on torch/_tensor_str.py:
```diff
@@ -225,7 +225,7 @@ def _str(self):
         size_str = str(tuple(self.shape)).replace(' ', '')
         return '{} of size {} with indices:\n{}\nand values:\n{}'.format(
-            self.type(), size_str, self._indices(), self._values())
+            self.type(), size_str, torch._indices(self), torch._values(self))
```
Yep, thanks

goldsborough committed:
Undo changes for sparseDims and denseDims, rename _sparse_mask to sparse_mask

goldsborough committed:
Fix sparse_mask

goldsborough committed:
Fix documentation

goldsborough committed:
Merge branch 'master' into remove-tensor-underscore
